File: server/routes/files.ts
import express from 'express';
import multer from 'multer';
import { authenticate } from '../middleware/auth';
import {
    listFiles,
    uploadFile,
    downloadFile,
    deleteFile,
    createFolder,
    readFile,
    updateFile,
    renameFile,
    extractZip,
    copyFiles,
    moveFiles,
    gitClone,
    changePermissions
} from '../controllers/fileController';

const router = express.Router();

// Configure multer for file uploads
const storage = multer.memoryStorage();
const upload = multer({
    storage,
    limits: {
        fileSize: 50 * 1024 * 1024 // 50MB max file size
    }
});

// All routes require authentication
router.use(authenticate);

// List files in directory
router.get('/list', listFiles);

// Upload file
router.post('/upload', upload.single('file'), uploadFile);

// Download file
router.get('/download', downloadFile);

// Read file content (for editor)
router.get('/read', readFile);

// Update file content (for editor)
router.put('/update', updateFile);

// Rename file/folder
router.put('/rename', renameFile);

// Delete file/folder
router.delete('/delete', deleteFile);

// Create folder
router.post('/folder', createFolder);

// Extract ZIP file
router.post('/extract', extractZip);

// Copy files/folders
router.post('/copy', copyFiles);

// Move files/folders
router.post('/move', moveFiles);

// Git Clone
router.post('/git-clone', authenticate, gitClone);
router.post('/permissions', authenticate, changePermissions);

export default router;

File: go-server/routes/v1/files.go
package v1

import (
	"cloudku-server/controllers"
	"cloudku-server/middleware"

	"github.com/gin-gonic/gin"
)

// RegisterFileRoutes sets up file manager routes
//
// # All routes require authentication - file operations are sensitive
//
// ENDPOINTS:
//   - GET    /files/list        - List files in directory
//   - GET    /files/stats       - Get storage statistics
//   - POST   /files/upload      - Upload file
//   - GET    /files/download    - Download file
//   - DELETE /files/delete      - Delete file/folder
//   - POST   /files/folder      - Create folder
//   - GET    /files/read        - Read file content
//   - PUT    /files/update      - Update file content
//   - PUT    /files/rename      - Rename file/folder
//   - POST   /files/copy        - Copy files
//   - POST   /files/move        - Move files
//   - POST   /files/extract     - Extract ZIP archive
//   - POST   /files/compress    - Compress to ZIP
//   - POST   /files/git-clone   - Clone Git repository
//   - PUT    /files/permissions - Change file permissions
func RegisterFileRoutes(rg *gin.RouterGroup, ctrl *controllers.FileController) {
	files := rg.Group("/files")
	files.Use(middleware.AuthMiddleware())
	{
		// List & Stats
		files.GET("/list", ctrl.ListFiles)
		files.GET("/stats", ctrl.GetStats)

		// CRUD Operations
		files.POST("/upload", ctrl.UploadFile)
		files.GET("/download", ctrl.DownloadFile)
		files.DELETE("/delete", ctrl.DeleteFile)
		files.POST("/folder", ctrl.CreateFolder)

		// Read & Write Operations
		files.GET("/read", ctrl.ReadFile)
		files.PUT("/update", ctrl.UpdateFile)
		files.PUT("/rename", ctrl.RenameFile)

		// Batch Operations
		files.POST("/copy", ctrl.CopyFiles)
		files.POST("/move", ctrl.MoveFiles)

		// Archive Operations
		files.POST("/extract", ctrl.ExtractZip)
		files.POST("/compress", ctrl.CompressFiles)

		// Git Integration
		files.POST("/git-clone", ctrl.GitClone)

		// Permissions Management
		files.PUT("/permissions", ctrl.ChangePermissions)
	}
}
